lib/triton_kernel_call.cc

```diff
@@ -0,0 +1,124 @@
+    CU_LAUNCH_PARAM_BUFFER_SIZE, &params_size,
+    CU_LAUNCH_PARAM_END
+  };
+  CUresult result = cuLaunchKernel(kernel, grid_0, grid_1, grid_2, num_warps * 32, 1, 1, shared_mem, stream, nullptr, config);
+  if (result != 0) {
+    std::cout << "Failed launch: " << result << std::endl;
+  }
+};
+
+void TritonExecutable::load(CUdevice device) {
+  // Mimics Triton kernel loading
+  std::string assembly;
+  if (asm_map.find("cubin") != asm_map.end())
+    assembly = py::cast<std::string>(asm_map["cubin"]);
+  else
+    assembly = py::cast<std::string>(asm_map["ptx"]);
+  CUfunction fun;
+  CUmodule mod;
+  cuModuleLoadData(&mod, assembly.c_str());
+  cuModuleGetFunction(&fun, mod, name.c_str());
+  int n_regs = 0;
+  int n_spills = 0;
+  cuFuncGetAttribute(&n_regs, CU_FUNC_ATTRIBUTE_NUM_REGS, fun);
+  cuFuncGetAttribute(&n_spills, CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES, fun);
+  n_spills /= 4;
+  int shared_optin;
+  cuDeviceGetAttribute(&shared_optin,
+                       CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK_OPTIN,
+                       device);
+  if (shared_mem > TRITON_MAX_N_SHARED_BYTES &&
+      shared_optin > TRITON_MAX_SHARED_OPTIN) {
+    cuFuncSetCacheConfig(fun, CU_FUNC_CACHE_PREFER_SHARED);
+    int shared_total, shared_static;
+    cuDeviceGetAttribute(
+        &shared_total, CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_MULTIPROCESSOR,
+        device);
+    cuFuncGetAttribute(&shared_static, CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES,
+                       fun);
+    cuFuncSetAttribute(fun, CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES,
+                       shared_optin - shared_static);
+  }
+  kernels[device] = fun;
+};
+
+void do_custom_call(CUstream stream, void** buffers,
+    char* opaque, size_t opaque_len) {
+  uint64_t descriptor = std::strtoull(opaque, NULL, 0);
+  TritonExecutable* executable = TritonExecutable::from_descriptor(descriptor);
+  executable->launch(stream, buffers);
+}
+
+std::pair<std::string, py::object> MakeTritonExecutable(std::string name, asm_map_t asm_map, uint32_t shared_mem, uint32_t grid_0, uint32_t grid_1, uint32_t grid_2, uint32_t num_warps, uint32_t arity) {
+  auto triton_call = std::make_unique<TritonExecutable>(
+      name, asm_map, shared_mem, grid_0, grid_1, grid_2, num_warps, arity);
+  std::string descriptor = std::to_string((uint64_t) triton_call.get());
+  py::capsule callback_capsule(triton_call.release(), [](void* ptr) {
+    delete reinterpret_cast<TritonExecutable*>(ptr);
+  });
+  return std::make_pair(descriptor, py::object(std::move(callback_capsule)));
+}
+
+template <typename T>
+pybind11::capsule EncapsulateFunction(T* fn) {
+  return pybind11::capsule(reinterpret_cast<void*>(fn), "xla._CUSTOM_CALL_TARGET");
+}
+
+PYBIND11_MODULE(triton_kernel_call, m) {
+  m.def("make_triton_call_descriptor", [](std::string name, asm_map_t asm_map, uint32_t shared_mem, uint32_t grid_0, uint32_t grid_1, uint32_t grid_2, uint32_t num_warps, uint32_t arity){
```

Comment:
It would be cleaner using function pointer:

m.def("make_triton_call_descriptor", &MakeTritonExecutable);

Reply:
Done.